components/server/src/projects/scm-service.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { Project, User } from "@gitpod/gitpod-protocol";
+import { RepoURL } from "../repohost";
+import { inject, injectable } from "inversify";
+import { HostContextProvider } from "../auth/host-context-provider";
+import { Config } from "../config";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+
+@injectable()
+export class ScmService {
+    constructor(
+        @inject(HostContextProvider) private readonly hostContextProvider: HostContextProvider,
+        @inject(Config) private readonly config: Config,
+    ) {}
+
+    async canInstallWebhook(currentUser: User, cloneURL: string) {
+        try {
+            const parsedUrl = RepoURL.parseRepoUrl(cloneURL);
+            const hostContext = parsedUrl?.host ? this.hostContextProvider.get(parsedUrl?.host) : undefined;
+            const authProvider = hostContext && hostContext.authProvider.info;
+            const type = authProvider && authProvider.authProviderType;
+            const host = authProvider?.host;
+            if (!type || !host) {
+                throw Error("Unknown host: " + parsedUrl?.host);
+            }
+            if (
+                type === "GitLab" ||
+                type === "Bitbucket" ||
+                type === "BitbucketServer" ||
+                (type === "GitHub" && (host !== "github.com" || !this.config.githubApp?.enabled))
+            ) {
+                const repositoryService = hostContext?.services?.repositoryService;
+                if (repositoryService) {
+                    return await repositoryService.canInstallAutomatedPrebuilds(currentUser, cloneURL);
+                }
+            }
+            // The GitHub App case isn't handled here due to a circular dependency problem.
+        } catch (error) {
+            log.error("Failed to check precondition for creating a project.");
+        }
+        return false;
+    }
+
+    async installWebhookForPrebuilds(project: Project, installer: User) {
+        // Install the prebuilds webhook if possible
+        const { teamId, cloneUrl } = project;
+        const parsedUrl = RepoURL.parseRepoUrl(project.cloneUrl);
+        const hostContext = parsedUrl?.host ? this.hostContextProvider.get(parsedUrl?.host) : undefined;
+        const authProvider = hostContext && hostContext.authProvider.info;
+        const type = authProvider && authProvider.authProviderType;
+        if (
+            type === "GitLab" ||
+            type === "Bitbucket" ||
+            type === "BitbucketServer" ||
+            (type === "GitHub" && (authProvider?.host !== "github.com" || !this.config.githubApp?.enabled))
```

Comment:
Feels like we are checking for all types we support explicitly. I wonder for which cases this code is _not_ applicable...? :thinking:

Reply:
this one https://github.com/gitpod-io/gitpod/pull/18532/files#diff-d6fdb9bb15400f4fbba00db81b2ce6cb7f7bb566dd97b93b9fd4e3fc9a79f479R2689